src/zarr/store/memory.py

```diff
@@ -114,9 +114,15 @@ async def list_dir(self, prefix: str) -> AsyncGenerator[str, None]:
 
         if prefix == "":
             keys_unique = set(k.split("/")[0] for k in self._store_dict.keys())
-            for key in keys_unique:
-                yield key
         else:
-            for key in self._store_dict:
-                if key.startswith(prefix + "/") and key != prefix:
-                    yield key.removeprefix(prefix + "/").split("/")[0]
+            # Our dictionary doesn't contain directory markers, but we want to include
+            # a pseudo directory when there's a nested item and we're listing an
+            # intermediate level.
+            keys_unique = {
+                key.removeprefix(prefix + "/").split("/")[0]
+                for key in self._store_dict
+                if key.startswith(prefix + "/") and key != prefix
+            }
+
+        for key in keys_unique:
+            yield key
```

Comment:
```suggestion
        yield from keys_unique
```

Reply:
Surprisingly(?), you can't use `yield from` in async generators: https://peps.python.org/pep-0525/#asynchronous-yield-from

Reply:
doh! I knew that.